Pattern compiling

TPGamesNL opened the pull request:
### Description
Replaces `SkriptParser#parse_i` with a system that compiles patterns into `PatternElement`s, and uses those for matching.
Also fixes some broken patterns, that the current parser accepts but this one does not.

---
**Target Minecraft Versions:** any
**Requirements:** none
**Related Issues:** none

APickledWalrus commented on src/main/java/ch/njol/skript/patterns/PatternCompiler.java:
```diff
@@ -0,0 +1,197 @@
+					choicePatternElement.add(prevFirst != null ? prevFirst : getEmpty());
+				}
+				choicePatternElement.add(getEmpty());
+			} else if (c == '¦') {
```
What are your thoughts on adding ":" as an option for parse marks? This is a change skript-parser made that I think would be beneficial for us to introduce (at least as a secondary option).

TPGamesNL replied:
I think it'd be fine, but we do need to check if there are addons that use that character in syntax. It may be good to open an issue / discussion to get some feedback on this as well

TPGamesNL replied:
https://github.com/SkriptLang/Skript/discussions/4177

APickledWalrus commented on src/main/java/ch/njol/skript/patterns/InvalidPatternException.java:
```diff
@@ -0,0 +1,33 @@
+
+import org.eclipse.jdt.annotation.Nullable;
+
+public class InvalidPatternException extends IllegalArgumentException {
```
Do you think `MalformedPatternException` should be made public and used in place of this?

TPGamesNL replied:
I think I'd rather see MalformedPatternException be removed and replaced with this class instead, because it'd be better to have the exception in the same location (package) as the pattern compiler.

APickledWalrus replied:
What do you think about renaming this to MalformedPatternException then? Part of me does prefer that name lol - it doesn't really matter though

TPGamesNL replied:
Sure

APickledWalrus commented on src/main/java/ch/njol/skript/patterns/TypePatternElement.java:
```diff
@@ -0,0 +1,200 @@
+		int timeStart = s.indexOf("@");
+		if (timeStart != -1) {
+			time = Integer.parseInt(s.substring(timeStart + 1));
+			s = "" + s.substring(0, timeStart);
```
Is the `"" + ` necessary?

TPGamesNL replied:
Nope, I probably copied this part from the old parser

APickledWalrus commented on src/main/java/ch/njol/skript/patterns/TypePatternElement.java:
```diff
@@ -0,0 +1,200 @@
+		if (flagMask != ~0) {
+			if ((flagMask & SkriptParser.PARSE_LITERALS) != 0)
+				stringBuilder.append("~");
+			if ((flagMask & SkriptParser.PARSE_EXPRESSIONS) != 0)
```
Should this be an `else if`? You can't have both parsing types right?

APickledWalrus commented on src/main/java/ch/njol/skript/lang/SkriptParser.java:
```diff
@@ -1315,221 +1321,227 @@ private static int getGroupLevel(final String pattern, final int j) {
+	}
+
+
+//	@Nullable
```
Should this all just be removed instead of commented out?

APickledWalrus reviewed (changes requested):
Some quick questions/thoughts I had looking over this - overall, implementation looks good :)

APickledWalrus reviewed (approved):
LGTM with the changes

TPGamesNL commented:
I also added a parsing shortcut, which made the loading time of a test script I used 45% shorter

Mwexim commented on src/main/java/ch/njol/skript/patterns/SkriptPattern.java:
```diff
@@ -55,4 +68,24 @@ public String toString() {
+			}
+			next = next.next;
+		}
+		if (keywords.size() == 0)
+			return null;
+		return keywords.toArray(new String[0]);
```
Can't you just use a list and check if the list is empty? I mean this seems very cumbersome to me, and nullable arrays are not quite productive. Is there a reason for this approach? 

Other than that, great changes and I will probably use this keyword system in skript-parser as well!

Mwexim commented on src/main/java/ch/njol/skript/patterns/SkriptPattern.java:
```diff
@@ -55,4 +68,24 @@ public String toString() {
+			}
+			next = next.next;
+		}
+		if (keywords.size() == 0)
```
Can't you just use a list and check if the list is empty? I mean this seems very cumbersome to me, and nullable arrays are not quite productive. Is there a reason for this approach? 

Other than that, great changes and I will probably use this keyword system in skript-parser as well!

TPGamesNL replied:
Oh I accidentally kept that in: because the first draft of that system was with a single keyword, which was Nullable.
I've changed it now

TPGamesNL commented:
#4177 

I've implemented parse tags, which are like parse marks, but they're a list of strings instead of a single integer. They can only be used with the character `:`, and parse marks can also use this character instead of `¦` (the broken pipe character still works for parse marks, but cannot be used for parse tags). You can use `ParseResult#tags` or `ParseResult#hasTag(String)` to check which tags matched.
- `(A:option A|B:option B)` will add either `A` or `B` to the parse tags (case kept!).

I've also added the following, from a suggestion in the discussion linked above:
- `(:option A|:option B)` will add either `option a` or `option b` to the tags (lowercase!) because the specified tag was empty.
- `:(boots|leggings|chestplate|helmet)` will add either `boots`, `leggings`, `chestplate` or `helmet` to the tags, because the tag was empty.

bluelhf commented:
@TPGamesNL — Improving the Skript parser 1000 lines at a time ^ ^ゞ 

LGTM! :)

TPGamesNL commented:
(for addon devs referring to this PR) An example usage could be the following:

currently there is the syntax `[(1¦async)] [script] (load|init|enable)`. The init section has this code: `asyncAllowed = parseResult.mark == 1;`. 

If we wanted to use a parse tag instead of a parse mark, we'd use this syntax: `[(:async)] [script] (load|init|enable)` with this code in init: `asyncAllowed = parseResult.hasTag("async");`
